src/axom/core/detail/FlatTable.hpp

```diff
@@ -120,20 +120,73 @@ struct GroupBucket
     return InvalidSlot;
   }
 
-  void setBucket(int index, std::uint8_t hash) { metadata.buckets[index] = reduceHash(hash); }
+  template <bool Atomic = false>
+  AXOM_HOST_DEVICE void setBucket(int index, std::uint8_t hash)
+  {
+    std::uint8_t reduced_hash = reduceHash(hash);
+#if defined(AXOM_USE_RAJA)
+    if(Atomic)  // TODO: should be constexpr
+    {
+  #if defined(AXOM_USE_CUDA)
+      // CUDA workaround for lack of 8-bit atomicStore.
+      volatile std::uint8_t* bucket = &(metadata.buckets[index]);
+      *bucket = reduced_hash;
+  #else
+      RAJA::atomicStore<RAJA::auto_atomic>(&(metadata.buckets[index]), reduced_hash);
+  #endif
+      return;
+    }
+#endif
+    metadata.buckets[index] = reduced_hash;
+  }
 
   void clearBucket(int index) { metadata.buckets[index] = Empty; }
 
-  void setOverflow(std::uint8_t hash)
+  template <bool Atomic = false>
+  AXOM_HOST_DEVICE void setOverflow(std::uint8_t hash)
   {
     std::uint8_t hashOfwBit = 1 << (hash % 8);
+#if defined(AXOM_USE_RAJA)
+    if(Atomic)  // TODO: should be constexpr
+    {
+  #if defined(AXOM_USE_HIP) || defined(AXOM_USE_CUDA)
+      // Workaround for a lack of an atomicOr builtin for uint8_t.
+      GroupBucket pack_data {};
+      pack_data.metadata.ofw = hashOfwBit;
+      RAJA::atomicOr<RAJA::auto_atomic>(&data[0], pack_data.data[0]);
+  #else
+      RAJA::atomicOr<RAJA::auto_atomic>(&(metadata.ofw), hashOfwBit);
+  #endif
+      return;
+    }
+#endif
     metadata.ofw |= hashOfwBit;
   }
 
-  bool getMaybeOverflowed(std::uint8_t hash) const
+  template <bool Atomic = false>
+  AXOM_HOST_DEVICE bool getMaybeOverflowed(std::uint8_t hash) const
   {
     std::uint8_t hashOfwBit = 1 << (hash % 8);
-    return (metadata.ofw & hashOfwBit);
+    std::uint8_t curr_ofw;
+#if defined(AXOM_USE_RAJA)
+    if(Atomic)  // TODO: should be constexpr
+    {
+  #if defined(AXOM_USE_CUDA)
+      // CUDA workaround for lack of 8-bit atomicLoad.
+      curr_ofw = *const_cast<const volatile std::uint8_t*>(&(metadata.ofw));
+  #else
+      // TODO: why is the const_cast required? (RAJA issue?)
+      curr_ofw = RAJA::atomicLoad<RAJA::auto_atomic>(const_cast<std::uint8_t*>(&(metadata.ofw)));
```

Comment:
Can we add an `axom::atomicLoad` to `axom/core/execution/atomics.hpp` and use it here?